src/audio/dai-zephyr.c

```diff
@@ -250,13 +307,44 @@ static enum dma_cb_status dai_dma_cb(struct dai_data *dd, struct comp_dev *dev,
 
 	local_buf = buffer_acquire(dd->local_buffer);
 
-	if (dev->direction == SOF_IPC_STREAM_PLAYBACK)
-		ret = dma_buffer_copy_to(local_buf, dma_buf,
-					 dd->process, bytes);
-	else
-		ret = dma_buffer_copy_from(dma_buf, local_buf,
-					   dd->process, bytes);
+	if (list_is_empty(&dev->bsink_list)) {
+		// endpoint case, directly copy/consume and produce
+		if (dev->direction == SOF_IPC_STREAM_PLAYBACK)
+			ret = dma_buffer_copy_to(local_buf, dma_buf,
+						 dd->process, bytes);
+		else
+			ret = dma_buffer_copy_from(dma_buf, local_buf,
```

Comment:
It looks like this "else" processing here is unnecessary. In case of capture, dev->bsink_list should not be empty but should contain one or more regular sink buffers. That case is already handled below in a loop.

Reply:
keep it for now, I tried with remove this part, I saw some input/output error with capture:
arecord: pcm_read:2221: read error: Input/output error